packages/babel-helper-builder-react-jsx-experimental/src/index.js

```diff
@@ -367,47 +313,40 @@ You can set \`throwIfNamespace: false\` to bypass this warning.`,
     }
   }
 
-  function addAutoImports(path, state) {
-    const imports = getImportNames(path, state);
-    if (isModule(path)) {
-      // import {jsx} from "react";
-      // import {createElement} from "react";
-      const importMap = {};
-
-      imports.forEach(importName => {
-        if (!importMap[importName]) {
-          importMap[importName] = addNamed(
-            path,
-            importName,
-            getSource(state.source, importName),
-            {
-              importedInterop: "uncompiled",
-              ensureLiveReference: true,
-            },
-          ).name;
-        }
-      });
+  function createImportLazily(pass, path, importName, source) {
+    return () => {
+      const actualSource = getSource(source, importName);
+      if (isModule(path)) {
+        let reference = pass.get(
+          `@babel/plugin-react-jsx/imports/${importName}`,
+        );
+        if (reference) return t.cloneNode(reference);
 
-      return importMap;
-    } else {
-      const importMap = {};
-      const sourceMap = {};
-      imports.forEach(importName => {
-        const source = getSource(state.source, importName);
-        if (!importMap[importName]) {
-          if (!sourceMap[source]) {
-            // var _react = require("react")
-            sourceMap[source] = addNamespace(path, source, {
-              importedInterop: "uncompiled",
-              ensureLiveReference: true,
-            }).name;
-          }
+        reference = addNamed(path, importName, actualSource, {
+          importedInterop: "uncompiled",
```

Comment:
previously `ensureLiveReference` was used for both call sites, are u confident that this can be dropped?

Reply:
Yes, because it's only needed for "named" imports in CJS modules: it injects
```js
var _mod = require("foo");
_mod.namedImport;
```
rather than
```js
var namedImport = require("foo").namedImport;
```

However, we only inject named imports in ESM modules (which are already "live") or namespace imports (aka plain `require` calls) in CJS.